### PR TITLE
Modify linkifyjs LinkifyOptions interface to allow extra return type for 'format'

### PR DESCRIPTION
**Pull request by cheshire137**
I noticed that I'm able to pass a `format` function that returns either a React component or a string, and Linkify renders either one. This modifies the type declaration for the `format` function so you can return React components in addition to strings. See https://soapbox.github.io/linkifyjs/docs/2.1.html#reactjs-interface for info about the React interface.

### CODE REVIEW

**Comment by cheshire137** [comment]
cc @szhu and @ovidiubute 🙇

File: types/linkifyjs/linkifyjs-tests.tsx [hunk 1]
```diff
@@ -68,6 +68,19 @@ describe("linkifyjs/html", () => {
...
+            if (type === 'url') {
+                return <span>{value}</span>;
+            }
+            return (
+                <span>
+                    {value} ({type})
+                </span>
+            );
+        },
+    });
```
Comment (szhu) [inline types/linkifyjs/linkifyjs-tests.tsx]: This doesn't seem to test that `format()` can return a string.

File: types/linkifyjs/index.d.ts [hunk 2]
```diff
@@ -94,7 +94,7 @@ export interface LinkifyOptions {
      * NOTE: According to the linkifyjs implementation, `format` can be just a
      * string, but this is not mentioned in the docs, so we exclude it.
```
Comment (szhu) [inline types/linkifyjs/index.d.ts]:
```suggestion
      * string, so we include it (even though this is not mentioned in the docs).
```

But actually maybe move this sentence to just above where `FormatFunc` is defined?

File: types/linkifyjs/index.d.ts [hunk 3]
```diff
@@ -13,7 +13,9 @@ export type PossiblyFuncOfHrefAndType<T> =
 
 // There's always a possibility that values at a string key can always be
 // undefined; i.e., if the key does not exist.
-export type PossiblyByType<T> = T | { [type: string]: T | undefined };
+export type PossiblyByType<T> = T | { [type: string]: T | React.ReactElement | undefined };
+
+export type FormatFunc = (value: string, type: string) => string | React.ReactElement;
```
Comment (szhu) [inline types/linkifyjs/index.d.ts]: Can you test whether returning `undefined`/`null`/`number`/an array of the above is allowed? If so, the return type can simply be `React.ReactNode`.

**Review by szhu — CHANGES_REQUESTED** [review]

**Review by szhu — COMMENTED** [review]
Thanks for contributing this change! Feel free to let me know if you have any questions

File: types/linkifyjs/linkifyjs-tests.tsx (hunk 1 above)
Comment (cheshire137) [reply types/linkifyjs/linkifyjs-tests.tsx]: There's an existing `format` test above that returns only strings, so I had this one returning only elements. I can also add a test where `format` returns strings and elements.

File: types/linkifyjs/index.d.ts (hunk 3 above)
Comment (cheshire137) [reply types/linkifyjs/index.d.ts]: Good idea! I just tried it, using `export type FormatFunc = (value: string, type: string) => React.ReactNode;` in my type declaration in my project. Returning `null` or `undefined` works and results in an `<a>` tag with no children, as you might expect. Returning a number works and results in that number being the clickable text of the link. Returning an array also works; I tried returning `['cat', 'dog']` and the link text became `catdog`.

File: types/linkifyjs/index.d.ts [hunk 4]
```diff
@@ -13,7 +13,11 @@ export type PossiblyFuncOfHrefAndType<T> =
 
 // There's always a possibility that values at a string key can always be
 // undefined; i.e., if the key does not exist.
-export type PossiblyByType<T> = T | { [type: string]: T | undefined };
+export type PossiblyByType<T> = T | { [type: string]: T | React.ReactElement | undefined };
```
Comment (szhu) [inline types/linkifyjs/index.d.ts]: What was this modification for? `PossiblyByType` is sort of a type that deals mostly with logic. I don't think `React.ReactElement` belongs here.

If there's a specific reason you were trying to use `React.ReactElement` for but you're not sure how to do it, feel free to explain and I can see if I can come up with a way to solve it as well.

File: types/linkifyjs/linkifyjs-tests.tsx (hunk 1 above)
Comment (szhu) [reply types/linkifyjs/linkifyjs-tests.tsx]: Oh, I totally forgot that `value` is a string -- that's why I didn't think any of the functions above were returning strings. Thanks for explaining!

**Comment by orta** [comment]
👋 - just going through old PRs, I think these changes are now integrated in the current types for linkify. Want to double check and close this if so?

**Comment by cheshire137** [comment]
Thank you, @orta!

**Pull request closed without merging**